cpp/src/TimerThread.cpp

```diff
@@ -0,0 +1,158 @@
+void TimerThread::TimerThreadEntryPoint
+(
+		Event* _exitEvent,
+		void* _context
+)
+{
+	TimerThread* timer = (TimerThread*)_context;
+	if( timer )
+	{
+		timer->TimerThreadProc( _exitEvent );
+	}
+}
+
+//-----------------------------------------------------------------------------
+// <TimerThread::TimerThread>
+// Constructor.
+//-----------------------------------------------------------------------------
+TimerThread::TimerThread
+(
+):
+m_timerEvent( new Event() ),
+m_timerMutex( new Mutex() ),
+m_timerTimeout( Wait::Timeout_Infinite )
+{
+}
+
+//-----------------------------------------------------------------------------
+// <TimerThread::~TimerThread>
+// Destructor.
+//-----------------------------------------------------------------------------
+TimerThread::~TimerThread
+(
+)
+{
+	m_timerMutex->Lock();
+	for ( list<TimerEventEntry *>::iterator it = m_timerEventList.begin(); it != m_timerEventList.end(); ++it ) {
+		delete (*it);
+	}
+	m_timerMutex->Unlock();
+	m_timerMutex->Release();
+	m_timerEvent->Release();
+}
+
+
+//-----------------------------------------------------------------------------
+// <TimerThread::TimerThreadProc>
+// Thread for timer based actions
+//-----------------------------------------------------------------------------
+void TimerThread::TimerThreadProc
+(
+		Event* _exitEvent
+)
+{
+	Log::Write( LogLevel_Info, "Timer: thread starting" );
+
+	Wait* waitObjects[2];
+	waitObjects[0] = _exitEvent;
+	waitObjects[1] = m_timerEvent;
+	uint32 count = 2;
+
+	// Initially no timer events so infinite timeout.
+	m_timerTimeout = Wait::Timeout_Infinite;
+
+	while( 1 )
+	{
+		Log::Write( LogLevel_Detail, "Timer: waiting with timeout %d", m_timerTimeout );
+		int32 res = Wait::Multiple( waitObjects, count, m_timerTimeout );
+
+		if (res == 0)
+		{
+			// Exit has been signalled
+			return;
+
+		} else {
+			// Timeout or new entry to timer list.
+			m_timerTimeout = Wait::Timeout_Infinite;
+
+			// Go through all waiting actions, and see if any need to be performed.
+			m_timerMutex->Lock();
+			list<TimerEventEntry *>::iterator it = m_timerEventList.begin();
+			while( it != m_timerEventList.end() ) {
+				int32 tr = (*it)->timestamp.TimeRemaining();
+				if (tr <= 0) {
+					// Expired so perform action and remove from list.
+					Log::Write( LogLevel_Info, "Timer: delayed event" );
+					(*it)->callback();
+					delete (*it);
+					it = m_timerEventList.erase( it );
+
+				} else {
+					// Time remaining.
+					m_timerTimeout = (m_timerTimeout == Wait::Timeout_Infinite) ? tr : std::min(m_timerTimeout, tr);
+					++it;
+				}
+			}
+			m_timerEvent->Reset();
+			m_timerMutex->Unlock();
+		}
+	} // while( 1 )
+}
+
+//-----------------------------------------------------------------------------
+// <TimerThread::TimerSetEvent>
+//-----------------------------------------------------------------------------
+void TimerThread::TimerSetEvent
+(
+		int32 _milliseconds,
+		TimerCallback _callback
+)
```

Comment:
may also want to pass a context void * pointer in, so for callbacks we can pass arbitrary struct etc (not needed for the delayed sending, but for future support)

Reply:
Just wanted to double check what was wanted here. The callback function can have its arguments set through std::bind to pass information in that way. Do you mean an extra pointer to store in a struct TimerEventEntry?

Reply:
ehhh. I missed that. 👍